src/Algebra/Graph/Acyclic/AdjacencyMap.hs

```diff
@@ -0,0 +1,112 @@
+module Algebra.Graph.Acyclic.AdjacencyMap where
+
+import qualified Algebra.Graph.AdjacencyMap as AM
+import qualified Algebra.Graph.AdjacencyMap.Algorithm as AMA
+import qualified Algebra.Graph.AdjacencyMap.Internal as AMI
+import qualified Algebra.Graph.NonEmpty.AdjacencyMap as NonEmpty
+import Control.Monad.Trans.Class (lift)
+import Control.Monad.Trans.State.Strict
+import Data.Map (Map, (!))
+import qualified Data.Map.Strict as Map
+import qualified Data.Set as Set
+
+newtype AdjacencyMap a = AAM
+  { aam :: AM.AdjacencyMap a
+  } deriving (Show)
+
+consistent :: Ord a => AdjacencyMap a -> Bool
+consistent (AAM m) = AMI.consistent m && AMA.isAcyclic m
+
+type AdjacencyMapP a = AdjacencyMap (Int, a)
+
+empty :: AdjacencyMap a
+empty = AAM AM.empty
+
+vertex :: a -> AdjacencyMap a
+vertex x = AAM $ AM.vertex x
+
+vertices :: (Ord a) => [a] -> AdjacencyMap a
+vertices xs = AAM $ AM.vertices xs
+
+overlayD ::
+     (Ord a, Ord b)
+  => AdjacencyMap a
+  -> AdjacencyMap b
+  -> AdjacencyMap (Either a b)
+overlayD (AAM a) (AAM b) = AAM (AM.overlay (AM.gmap Left a) (AM.gmap Right b))
+
+connectD ::
+     (Ord a, Ord b)
+  => AdjacencyMap a
+  -> AdjacencyMap b
+  -> AdjacencyMap (Either a b)
+connectD (AAM a) (AAM b) = AAM (AM.connect (AM.gmap Left a) (AM.gmap Right b))
+
+transitiveClosure :: Ord a => AdjacencyMap a -> AdjacencyMap a
+transitiveClosure (AAM x) = AAM (AM.transitiveClosure x)
+
+overlay :: (Ord a) => AdjacencyMap a -> AdjacencyMap a -> AdjacencyMap a
+overlay (AAM x) (AAM y) = AAM (AM.overlay x y)
+
+connect :: (Ord a) => AdjacencyMap a -> AdjacencyMap a -> AdjacencyMap a
+connect (AAM x) (AAM y) =
+  AAM . AMI.AM $
+  Map.unionsWith Set.union $
+  AM.adjacencyMap x :
+  AM.adjacencyMap y :
+  [ Map.fromSet
+      (\k -> Set.filter (< k) . Map.keysSet $ AM.adjacencyMap y)
+      (Map.keysSet $ AM.adjacencyMap x)
+  ]
+
+-- reverse and Map.keys results in an ascending list
+topSort :: (Ord a) => AdjacencyMap a -> [a]
```

Comment:
This is incorrect: you need to call the `AM.topSort` function, because the topological sort does not necessarily have anything to do with the order on keys. Try to test this function using `isTopSort` and you'll see that the test will fail. 

As a general comment, we should test all functions.

Reply:
There is an invariant that I have in mind:
**Invariant:** The edges in the graph only depend on the order on its vertices. There can only be edges from a higher ordered vertex to a lower ordered vertex. All the functions in this PR are made keeping that in mind.
Keeping the invariant in mind it would result in a correct topological ordering.

Reply:
But we use `consistent` to capture all our invariants, and at the moment it doesn't include this extra complexity. As I mentioned, if you use the function `AM.topSort` you do not need to rely on any extra invariants: `consistent` provides enough guarantees. Why do you need this extra complexity, which requirement does it help to achieve?

Reply:
For simplicity in adding the edges in the future. But maybe I'm thinking too far ahead. I think it's better to build on simple building blocks as you suggested :-)